ambry-store/src/main/java/com/github/ambry/store/HybridCompactionPolicy.java

```diff
@@ -0,0 +1,191 @@
+      return new StatsBasedCompactionPolicy(storeConfig, time);
+    }
+    if (readyToTriggerCompactionAllPolicy(compactionPolicySwitchInfo)) {
+      logger.trace("Return CompactAllPolicy this round");
+      updateCompactionInfoWhenCompactAll(compactionPolicySwitchInfo);
+      return new CompactAllPolicy(storeConfig, time);
+    } else {
+      if (compactionPolicySwitchInfo.getCompactionPolicyCounter() == null) {
+        logger.trace("Counter is null");
+      } else {
+        logger.trace("Return StatsBasedCompactionPolicy this round");
+      }
+      return new StatsBasedCompactionPolicy(storeConfig, time);
+    }
+  }
+
+  /**
+   * Determine which compactionPolicy to use for current compaction cycle.
+   * @param compactionPolicySwitchInfo the info to determine which {@link CompactionPolicy} to use this round.
+   * @return {@code true} if the counter value equals to 0 or it's storeCompactionPolicySwitchPeriod days past the start time of CompactAllPolicy.
+   */
+  private boolean readyToTriggerCompactionAllPolicy(CompactionPolicySwitchInfo compactionPolicySwitchInfo) {
+    return compactionPolicySwitchInfo.getCompactionPolicyCounter() != null
+        && compactionPolicySwitchInfo.getCompactionPolicyCounter().getCounter() == 0 ||
+        compactionPolicySwitchInfo.getLastCompactAllTime() + TimeUnit.DAYS.toMillis(
+            storeConfig.storeCompactionPolicySwitchTimestampDays) <= System.currentTimeMillis();
+  }
+
+  /**
+   * Update the {@link CompactionPolicySwitchInfo} before the start of {@link CompactAllPolicy}
+   * Once the compactAllPolicy has been triggered, no matter it's been triggered by timestamp or counter value
+   * the lastCompactAllTime will be set to current time and the counter value will reset to 0.
+   * @param compactionPolicySwitchInfo the info to determine which {@link CompactionPolicy} to use this round.
+   */
+  private void updateCompactionInfoWhenCompactAll(CompactionPolicySwitchInfo compactionPolicySwitchInfo) {
+    compactionPolicySwitchInfo.setLastCompactAllTime(System.currentTimeMillis());
+    compactionPolicySwitchInfo.getCompactionPolicyCounter().setCounter(INIT_COUNTER_VALUE);
+  }
+
+  /**
+   * @return blobToCompactionPolicySwitchInfoMap which key is storeId and value is {@link CompactionPolicySwitchInfo}
+   */
+  Map<String, CompactionPolicySwitchInfo> getBlobToCompactionPolicySwitchInfoMap() {
+    return this.blobToCompactionPolicySwitchInfoMap;
+  }
+
+  /**
+   * Back up {@link CompactionPolicySwitchInfo} in Json format for certain {@link BlobStore}
+   * @param dataDir The directory to store the file.
+   * @param compactionPolicySwitchInfo the info to determine which {@link CompactionPolicy} to use this round.
+   */
+  private void backUpCompactionPolicyInfo(String dataDir, CompactionPolicySwitchInfo compactionPolicySwitchInfo) {
+    if (dataDir != null && !dataDir.isEmpty()) {
+      File tempFile = new File(Paths.get(dataDir, COMPACT_POLICY_INFO_PATH).toString());
```

Comment:
nit: I encourage to create and write data to a temporary file and then rename the temporary file to COMPACT_POLICY_INFO_PATH file
Something like 
```
File tempFile = new File(dataDir, COMPACT_POLICY_INFO_PATH + ".temp");
if (!tempFile.exists()) {
    tempFile.createNewFile();
}
objectMapper.defaultPrettyPrintingWriter().writeValue(tempFile, compactionPolicySwitchInfo);
tempFile.renameTo(new File(dataDir, COMPACT_POLICY_INFO_PATH);
```
The reason to use a temporary file and rename it after is that POSIX guarantees file rename is an atomic operation, but not the file write. The json file is pretty small so that file writing should be atomic, but it's just a good pattern to follow.

Reply:
Updated to use temp file and rename after write. I didn't add the file exist check since createNewFile() is able to handle it.